### ISSUE TITLE
drawMatches output transparent when drawing on 4 channel image

### ISSUE THREAD

**Issue by dan-masek** [opencv/opencv#13219]
Discovered in https://stackoverflow.com/questions/53383569/cv2-drawmatches-draws-on-blank-screen

Judging from the source code, relevant for all branches.

----

Using `drawMatches` on 4 channel image with random colours causes the output to be transparent.

Culprits are clear:
* https://github.com/opencv/opencv/blob/master/modules/features2d/src/draw.cpp#L120
* https://github.com/opencv/opencv/blob/master/modules/features2d/src/draw.cpp#L176

`Scalar color = isRandMatchColor ? Scalar( rng(256), rng(256), rng(256) ) : matchColor;`

In both cases the random colour generation leaves the 4th component of the `Scalar` at the default value of 0 (i.e. fully transparent in case of alpha).

Resolution seems simple, generate the colour as `Scalar( rng(256), rng(256), rng(256), 255 )`

**Comment by dan-masek** [comment]
Script that reproduces the problem (using images from OpenCV samples):

    import numpy as np
    import cv2

    imageLeft = cv2.imread('box.png')
    imageRight = cv2.imread('box_in_scene.png')


    orb = cv2.ORB_create()
    kpLeft, desLeft = orb.detectAndCompute(imageLeft, None)
    kpRight, desRight = orb.detectAndCompute(imageRight, None)

    bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)
    matches = bf.match(desLeft, desRight)

    goodMatches = []
    for m in matches:
        if m.distance < 100:
            goodMatches.append(m)
            
    output = np.zeros((max(imageLeft.shape[0], imageRight.shape[0]), imageLeft.shape[1] + imageRight.shape[1], 4), np.uint8)

    cv2.drawMatches(imageLeft, kpLeft, imageRight, kpRight, [goodMatches[0]], output, flags=cv2.DRAW_MATCHES_FLAGS_DRAW_OVER_OUTIMG)

    cv2.imwrite('output.png', output)

Output:

![output_old](https://user-images.githubusercontent.com/10824015/48781663-bcd78000-ecdc-11e8-84dc-631626ab4e16.png)

Output with #13223 applied:

![output_fixed](https://user-images.githubusercontent.com/10824015/48781629-af21fa80-ecdc-11e8-91d1-cafcaecdb129.png)

**Comment by hsawano** [comment]
# Comment for OpenCV Issue #13219

## iOS CvVideoCamera Integration: Additional Root Cause and Solution Pattern

### Background

The issue of transparent `drawMatches()` output on 4-channel images (#13219 / #13223) has been resolved in the OpenCV codebase by ensuring the alpha channel is explicitly set to 255 in random color generation.

However, I've discovered an **additional and related root cause** that affects iOS developers using `CvVideoCamera`, which stems from a **memory management boundary problem** rather than OpenCV itself.

### The Additional Problem: Reference Parameter Memory Management

When integrating OpenCV with iOS AVFoundation through `CvVideoCamera`, a similar transparency issue can occur due to conflicting memory management designs:

**Symptom:**
```objc
void processImage(cv::Mat& image) {
    cv::Mat result_bgr;
    cv::drawMatches(..., result_bgr, ...);
    
    cv::Mat result_bgra;
    cv::cvtColor(result_bgr, result_bgra, cv::COLOR_BGR2BGRA);
    
    std::vector<cv::Mat> channels;
    cv::split(result_bgra, channels);
    channels[3] = cv::Mat(result_bgra.size(), CV_8UC1, cv::Scalar(255));
    cv::merge(channels, image);  // ❌ Result becomes transparent!
}
```

### Root Cause: Memory Management Conflict

**OpenCV's Design Assumption:**
- Functions like `cv::merge(vector<Mat> src, Mat& dst)` assume **complete ownership** of the output parameter
- Memory initialization handled internally

**iOS AVFoundation's Design:**
- `CvVideoCamera.processImage(cv::Mat& image)` provides **reference parameters** to camera frame buffers
- Memory ownership remains with AVFoundation
- Reference parameter points to pre-allocated camera frame buffer memory

**The Conflict:**
```
Reference Parameter (AVFoundation-owned memory)
    ↓
cv::merge(channels, image);
    ↓
OpenCV: "I will initialize this memory"
AVFoundation: "I already own this memory"
    ↓
Result: Incomplete initialization → Alpha channel undefined
```

### Solution: Intermediate Variable Pattern

**✅ Correct approach:**

```cpp
void processImage(cv::Mat& image) {
    cv::Mat result_bgr;
    cv::drawMatches(..., result_bgr, ...);
    
    cv::Mat result_bgra;
    cv::cvtColor(result_bgr, result_bgra, cv::COLOR_BGR2BGRA);
    
    std::vector<cv::Mat> channels;
    cv::split(result_bgra, channels);
    channels[3] = cv::Mat(result_bgra.size(), CV_8UC1, cv::Scalar(255));
    cv::merge(channels, result_bgra);  // ✅ Merge to intermediate variable
    
    result_bgra.copyTo(image);  // ✅ Explicit copy to reference parameter
}
```

**Why This Works:**
1. OpenCV achieves complete ownership of `result_bgra`
2. Alpha channel properly initialized to 255
3. Memory ownership separated: OpenCV handles `result_bgra`, AVFoundation handles `image`
4. `copyTo()` is a reliable low-level memory copy operation

### Generalization: iOS Framework Integration Pattern

This pattern applies whenever integrating C++ libraries with iOS native frameworks:

1. **ARKit CVPixelBuffer** - Use intermediate Mat for format conversion
2. **Vision Framework CGImage** - Avoid direct reference parameter modification
3. **Metal MTLTexture** - Explicit GPU ↔ CPU memory transfer

### Recommendation for OpenCV Documentation

This issue highlights a gap in OpenCV's iOS integration documentation:

1. **Add iOS Framework Integration Guide** - Best practices for reference parameter handling
2. **Document Reference Parameter Behavior** - Clarify memory ownership assumptions
3. **Provide Code Examples** - Show correct patterns for CvVideoCamera integration

### Status

- **Problem Type**: Memory management boundary issue (not a bug)
- **Severity**: High for iOS OpenCV developers
- **Solution**: Intermediate variable + copyTo pattern
- **Applicability**: General rule for framework integration

---

This supplementary root cause analysis should help iOS developers understand and resolve transparency issues beyond the random color generation fix addressed in #13223.
